File: core/src/main/kotlin/io/autocomplete/prompt/PromptTemplate.kt
package io.autocomplete.prompt

import io.autocomplete.client.ProviderKind

/**
 * How the completion request is serialized on the wire.
 */
enum class WireFormat {
    /** OpenAI-compatible FIM API: JSON fields `prompt` + `suffix` (Codestral, etc.). */
    FIM_FIELDS,

    /** OpenAI completions style: single `prompt` string (token FIM). */
    COMPLETION_PROMPT,

    /** OpenAI chat style: `messages` array. */
    CHAT_MESSAGES,
}

/**
 * Common completion prompt templates users can pick and probe.
 *
 * [AUTO] resolves at request time via [PromptTemplateDetector].
 */
enum class PromptTemplate {
    AUTO,
    CODESTRAL_API,
    QWEN,
    DEEPSEEK,
    STARCODER,
    CHAT,
    ;

    fun isAuto(): Boolean = this == AUTO

    fun wireFormat(): WireFormat =
        when (this) {
            AUTO -> WireFormat.CHAT_MESSAGES
            CODESTRAL_API -> WireFormat.FIM_FIELDS
            QWEN, DEEPSEEK, STARCODER -> WireFormat.COMPLETION_PROMPT
            CHAT -> WireFormat.CHAT_MESSAGES
        }

    /** Stable short label for logs / defaults (English). */
    fun shortLabel(): String =
        when (this) {
            AUTO -> "Auto"
            CODESTRAL_API -> "(fim) OpenAI FIM"
            QWEN -> "(fim) Qwen"
            DEEPSEEK -> "(fim) DeepSeek"
            STARCODER -> "(fim) StarCoder"
            CHAT -> "(chat) Pseudo-FIM"
        }

    fun stopTokens(): List<String> =
        when (this) {
            QWEN -> listOf("<|endoftext|>", "<|fim_prefix|>", "<|fim_middle|>", "<|fim_suffix|>", "<|fim_pad|>", "<|repo_name|>", "<|file_sep|>", "<|im_end|>")
            DEEPSEEK -> listOf("<｜fim▁begin｜>", "<｜fim▁hole｜>", "<｜fim▁end｜>", "<|EOT|>", "<｜end▁of▁sentence｜>")
            STARCODER -> listOf("<|endoftext|>", "<fim_prefix>", "<fim_middle>", "<fim_suffix>", "<fim_pad>", "<file_sep>", "<|eos|>")
            else -> emptyList()
        }

    /**
     * Build the model-facing prompt text for token-style templates.
     * For [CODESTRAL_API] / [CHAT] the client uses structured fields instead.
     */
    fun formatTokenPrompt(
        prefix: String,
        suffix: String,
    ): String =
        when (this) {
            QWEN -> "<|fim_prefix|>$prefix<|fim_suffix|>$suffix<|fim_middle|>"
            DEEPSEEK -> "<｜fim▁begin｜>$prefix<｜fim▁hole｜>$suffix<｜fim▁end｜>"
            STARCODER -> "<fim_prefix>$prefix<fim_suffix>$suffix<fim_middle>"
            AUTO, CODESTRAL_API, CHAT -> prefix
        }

    companion object {
        fun fromStored(value: String?): PromptTemplate {
            if (value.isNullOrBlank()) return AUTO
            return runCatching { valueOf(value.trim().uppercase()) }.getOrDefault(AUTO)
        }

        /** Concrete templates users can probe (excludes AUTO). */
        fun probeCandidates(): List<PromptTemplate> =
            listOf(CODESTRAL_API, QWEN, DEEPSEEK, STARCODER, CHAT)

        /**
         * Migrate legacy requestStyle setting into a template when promptTemplate is absent.
         */
        fun fromLegacyRequestStyle(requestStyle: String?): PromptTemplate =
            when (requestStyle?.uppercase()) {
                "FIM" -> CODESTRAL_API
                "CHAT" -> CHAT
                else -> AUTO
            }
    }
}

object PromptTemplateDetector {
    fun detect(
        model: String,
        providerKind: ProviderKind = ProviderKind.OPENAI_COMPATIBLE,
    ): PromptTemplate {
        if (providerKind == ProviderKind.MISTRAL_FIM) return PromptTemplate.CODESTRAL_API
        val m = model.trim().lowercase()
        if (m.isEmpty()) return PromptTemplate.CHAT
        return when {
            m.contains("codestral") || m.contains("mistral-code") || m.contains("devstral") ->
                PromptTemplate.CODESTRAL_API
            m.contains("deepseek") && (m.contains("coder") || m.contains("code")) ->
                PromptTemplate.DEEPSEEK
            m.contains("qwen") ->
                PromptTemplate.QWEN
            m.contains("codegemma") ->
                PromptTemplate.QWEN
            m.contains("starcoder") || m.contains("santacoder") || m.contains("codellama") ||
                m.contains("code-llama") || m.contains("crystalcoder") ->
                PromptTemplate.STARCODER
            m.contains("gpt-") || m.contains("claude") || m.contains("o1") || m.contains("o3") ||
                m.contains("o4") || m.contains("chatgpt") ->
                PromptTemplate.CHAT
            m.contains("coder") || m.contains("code-") || m.endsWith("-code") ->
                PromptTemplate.QWEN
            else -> PromptTemplate.CHAT
        }
    }

    fun resolve(
        stored: PromptTemplate,
        model: String,
        providerKind: ProviderKind,
    ): PromptTemplate = if (stored.isAuto()) detect(model, providerKind) else stored

    fun isRecognized(
        model: String,
        providerKind: ProviderKind = ProviderKind.OPENAI_COMPATIBLE,
    ): Boolean {
        if (model.isBlank()) return false
        val detected = detect(model, providerKind)
        // Unrecognized-ish: fell through to CHAT without chat-family keywords.
        if (detected != PromptTemplate.CHAT) return true
        val m = model.lowercase()
        return listOf("gpt-", "claude", "o1", "o3", "o4", "chatgpt", "instruct", "chat").any { m.contains(it) }
    }
}

enum class TemplateProbeStatus {
    SUCCESS,
    EMPTY,
    FAILED,
}

data class TemplateProbeResult(
    val template: PromptTemplate,
    val status: TemplateProbeStatus,
    val httpStatus: Int? = null,
    val latencyMs: Long = 0,
    val preview: String = "",
    val error: String = "",
    val resolvedPath: String = "",
) {
    fun isUsable(): Boolean = status == TemplateProbeStatus.SUCCESS
}

File: packages/core-ts/src/promptTemplate.ts
import type { PromptTemplateId, ProviderKind, WireFormat } from "./types.js";

export interface TemplateMeta {
  id: PromptTemplateId;
  shortLabel: string;
  wireFormat: WireFormat;
  stopTokens: string[];
}

const META: Record<Exclude<PromptTemplateId, "AUTO">, TemplateMeta> = {
  CODESTRAL_API: {
    id: "CODESTRAL_API",
    shortLabel: "(fim) OpenAI FIM",
    wireFormat: "FIM_FIELDS",
    stopTokens: [],
  },
  QWEN: {
    id: "QWEN",
    shortLabel: "(fim) Qwen",
    wireFormat: "COMPLETION_PROMPT",
    stopTokens: [
      "<|endoftext|>",
      "<|fim_prefix|>",
      "<|fim_middle|>",
      "<|fim_suffix|>",
      "<|fim_pad|>",
      "<|repo_name|>",
      "<|file_sep|>",
      "<|im_end|>",
    ],
  },
  DEEPSEEK: {
    id: "DEEPSEEK",
    shortLabel: "(fim) DeepSeek",
    wireFormat: "COMPLETION_PROMPT",
    stopTokens: [
      "<｜fim▁begin｜>",
      "<｜fim▁hole｜>",
      "<｜fim▁end｜>",
      "<|EOT|>",
      "<｜end▁of▁sentence｜>",
    ],
  },
  STARCODER: {
    id: "STARCODER",
    shortLabel: "(fim) StarCoder",
    wireFormat: "COMPLETION_PROMPT",
    stopTokens: [
      "<|endoftext|>",
      "<fim_prefix>",
      "<fim_middle>",
      "<fim_suffix>",
      "<fim_pad>",
      "<file_sep>",
      "<|eos|>",
    ],
  },
  CHAT: {
    id: "CHAT",
    shortLabel: "(chat) Pseudo-FIM",
    wireFormat: "CHAT_MESSAGES",
    stopTokens: [],
  },
};

export function fromStored(value: string | null | undefined): PromptTemplateId {
  if (!value?.trim()) return "AUTO";
  const u = value.trim().toUpperCase();
  if (u === "AUTO" || u in META) return u as PromptTemplateId;
  return "AUTO";
}

export function isAuto(t: PromptTemplateId): boolean {
  return t === "AUTO";
}

export function wireFormat(t: PromptTemplateId): WireFormat {
  if (t === "AUTO") return "CHAT_MESSAGES";
  return META[t].wireFormat;
}

export function shortLabel(t: PromptTemplateId): string {
  if (t === "AUTO") return "Auto";
  return META[t].shortLabel;
}

export function stopTokens(t: PromptTemplateId): string[] {
  if (t === "AUTO") return [];
  return META[t].stopTokens;
}

export function formatTokenPrompt(t: PromptTemplateId, prefix: string, suffix: string): string {
  switch (t) {
    case "QWEN":
      return `<|fim_prefix|>${prefix}<|fim_suffix|>${suffix}<|fim_middle|>`;
    case "DEEPSEEK":
      return `<｜fim▁begin｜>${prefix}<｜fim▁hole｜>${suffix}<｜fim▁end｜>`;
    case "STARCODER":
      return `<fim_prefix>${prefix}<fim_suffix>${suffix}<fim_middle>`;
    default:
      return prefix;
  }
}

export function probeCandidates(): PromptTemplateId[] {
  return ["CODESTRAL_API", "QWEN", "DEEPSEEK", "STARCODER", "CHAT"];
}

export function fromLegacyRequestStyle(requestStyle: string | null | undefined): PromptTemplateId {
  switch (requestStyle?.toUpperCase()) {
    case "FIM":
      return "CODESTRAL_API";
    case "CHAT":
      return "CHAT";
    default:
      return "AUTO";
  }
}

export function detectTemplate(
  model: string,
  providerKind: ProviderKind = "OPENAI_COMPATIBLE",
): PromptTemplateId {
  if (providerKind === "MISTRAL_FIM") return "CODESTRAL_API";
  const m = model.trim().toLowerCase();
  if (!m) return "CHAT";
  if (m.includes("codestral") || m.includes("mistral-code") || m.includes("devstral")) {
    return "CODESTRAL_API";
  }
  if (m.includes("deepseek") && (m.includes("coder") || m.includes("code"))) return "DEEPSEEK";
  if (m.includes("qwen") || m.includes("codegemma")) return "QWEN";
  if (
    m.includes("starcoder") ||
    m.includes("santacoder") ||
    m.includes("codellama") ||
    m.includes("code-llama") ||
    m.includes("crystalcoder")
  ) {
    return "STARCODER";
  }
  if (
    m.includes("gpt-") ||
    m.includes("claude") ||
    m.includes("o1") ||
    m.includes("o3") ||
    m.includes("o4") ||
    m.includes("chatgpt")
  ) {
    return "CHAT";
  }
  if (m.includes("coder") || m.includes("code-") || m.endsWith("-code")) return "QWEN";
  return "CHAT";
}

export function resolveTemplate(
  stored: PromptTemplateId,
  model: string,
  providerKind: ProviderKind,
): PromptTemplateId {
  return isAuto(stored) ? detectTemplate(model, providerKind) : stored;
}

export function isRecognized(
  model: string,
  providerKind: ProviderKind = "OPENAI_COMPATIBLE",
): boolean {
  if (!model.trim()) return false;
  const detected = detectTemplate(model, providerKind);
  if (detected !== "CHAT") return true;
  const m = model.toLowerCase();
  return ["gpt-", "claude", "o1", "o3", "o4", "chatgpt", "instruct", "chat"].some((k) =>
    m.includes(k),
  );
}
